File: creational/builder/python/classes/main_dish_builder.py
if __package__ is None or __package__ == '':
   # print(f'no package - init Interfaces')
   from interfaces import MealBuilderProtocol
   from meal_box import MealBox
   from meals import *
else:
   # print(f'package : {__package__} - init Interfaces')
   from .interfaces import MealBuilderProtocol
   from .meal_box import MealBox
   from .meals import *

class MainDishBuilder(MealBuilderProtocol):
   def __init__(self) -> None:
      self._meal : MealBox = MealBox()
   
   def reset(self) -> type:
      self._meal = MealBox()
      return self   
   
   def make_meal(self) -> type:
      RICE = Rice('Arroz', 5)
      BEANS = Beans('Feijao', 10)
      MEAT = Meat('Carne', 20)
      self._meal.add([RICE, BEANS, MEAT])
      return self

   def make_rice(self) -> type:
      RICE = Rice('Arroz', 6)
      self._meal.add(RICE)
      return self

   def make_beverage(self) -> type:
      BEVERAGE = Beverage('Bebida', 7)
      self._meal.add([BEVERAGE])
      return self
   
   def make_dessert(self) -> type: 
      DESSERT = Dessert('Sobremesa', 10)
      self._meal.add([DESSERT])
      return self
   
   def get_meal(self) -> MealBox:
      return self._meal
   
   def get_price(self) -> int:
      return self._meal.get_price()

File: src/creational/builder/typescript/classes/main-dish-builder.ts
import { MealBuilderProtocol } from './interfaces/meal-builder-protocol';
import { MealBox } from './meal-box';
import { Beans, Beverage, Dessert, Meat, Rice } from './meals';

export class MainDishBuilder implements MealBuilderProtocol {
  private _meal: MealBox = new MealBox();

  reset(): this {
    this._meal = new MealBox();
    return this;
  }

  getPrice(): number {
    return this._meal.getPrice();
  }

  getMeal(): MealBox {
    return this._meal;
  }

  makeRice(): this {
    const rice = new Rice('Arroz', 6);
    this._meal.add(rice);
    return this;
  }

  makeBeans(): this {
    const beans = new Beans('Feijão', 10);
    this._meal.add(beans);
    return this;
  }

  makeMeat(): this {
    const meat = new Meat('Carne', 12);
    this._meal.add(meat);
    return this;
  }

  makeMeal(): this {
    return this.makeRice().makeBeans().makeMeat();
  }
  makeBeverage(): this {
    const beverage = new Beverage('Bebida', 3);
    this._meal.add(beverage);
    return this;
  }
  makeDessert(): this {
    const dessert = new Dessert('Sobremesa', 12);
    this._meal.add(dessert);
    return this;
  }
}
